hibike/hibike_process.py

```diff
@@ -83,32 +61,54 @@ def hibike_process(badThingsQueue, stateQueue, pipeFromChild):
                 instruction_queue.put(("disable", []))            
 
 
-def device_write_thread(ser, queue):
+def device_write_thread(dev, queue):
     while True:
         instruction, args = queue.get()
 
+        # flush the queue if it's too big
+        if queue.qsize() > 10:
+            print("flushing queue, size is %d" % queue.qsize())
+            while instruction != "disable" and not queue.empty():
+                instruction, args = queue.get()
+
         if instruction == "ping":
-            hm.send(ser, hm.make_ping())
+            hm.hid_send(dev, hm.make_ping())
         elif instruction == "subscribe":
             uid, delay, params = args
-            hm.send(ser, hm.make_subscription_request(hm.uid_to_device_id(uid), params, delay))
+            hm.hid_send(dev, hm.make_subscription_request(hm.uid_to_device_id(uid), params, delay))
         elif instruction == "read":
             uid, params = args
-            hm.send(ser, hm.make_device_read(hm.uid_to_device_id(uid), params))
+            hm.hid_send(dev, hm.make_device_read(hm.uid_to_device_id(uid), params))
         elif instruction == "write":
             uid, params_and_values = args
-            hm.send(ser, hm.make_device_write(hm.uid_to_device_id(uid), params_and_values))
+            hm.hid_send(dev, hm.make_device_write(hm.uid_to_device_id(uid), params_and_values))
         elif instruction == "disable":
-            hm.send(ser, hm.make_disable())
+            hm.hid_send(dev, hm.make_disable())
         elif instruction == "heartResp":
             uid = args[0]
-            hm.send(ser, hm.make_heartbeat_response(hm.uid_to_device_id(uid)))
+            hm.hid_send(dev, hm.make_heartbeat_response())
+
+# global counters for tracking statistics
 
+# number of packets read
+ctr = 0
+start_time = time.time()
+# total time spent parsing received packets
+parse_time = 0
 
-def device_read_thread(index, ser, instructionQueue, errorQueue, stateQueue):
+def device_read_thread(index, dev, instructionQueue, errorQueue, stateQueue):
     uid = None
+    global ctr
```

Comment:
the slowdowns were definitely present before I added this to debug them. There are no locks on this global; at worst race conditions cause it to fail to increment at times, and at best python itself has some locking mechanism for this but the locks are released after very short increment operations.